File: app/src/main/java/com/ruideraj/secretelephant/main/MainViewModel.java
package com.ruideraj.secretelephant.main;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.ruideraj.secretelephant.AccountManager;
import com.ruideraj.secretelephant.R;
import com.ruideraj.secretelephant.SingleLiveEvent;

import javax.inject.Inject;

public class MainViewModel extends ViewModel implements AccountManager.AccountListener {

    public final MutableLiveData<Boolean> signedIn = new MutableLiveData<>();
    public final SingleLiveEvent<Integer> signOutMessage = new SingleLiveEvent<>();

    private AccountManager mAccountManager;

    @Inject
    public MainViewModel(AccountManager accountManager) {
        mAccountManager = accountManager;
        signedIn.setValue(false);
    }

    public void start() {
        boolean accountPresent = mAccountManager.getAccount() != null;
        if(signedIn.getValue() != accountPresent) signedIn.setValue(accountPresent);
    }

    @Override
    public void onSignOutSuccess() {
        signedIn.setValue(false);
        signOutMessage.setValue(R.string.main_menu_signed_out);
    }

    @Override
    public void onSignOutFailure() {
        signOutMessage.setValue(R.string.main_menu_sign_out_failed);
    }

    public void signOut() {
        mAccountManager.signOut(this);
    }
}

File: app/src/main/java/com/ruideraj/secretelephant/main/MainViewModel.kt
package com.ruideraj.secretelephant.main

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.ruideraj.secretelephant.AccountManager
import com.ruideraj.secretelephant.R
import com.ruideraj.secretelephant.SingleLiveEvent
import javax.inject.Inject

class MainViewModel @Inject constructor (private val accountManager: AccountManager): ViewModel() {

    val signedIn: LiveData<Boolean>
        get() = signedInData
    private val signedInData = MutableLiveData(false)

    val signOutMessage: LiveData<Int>
        get() = signOutMessageData
    private val signOutMessageData = SingleLiveEvent<Int>()

    fun start() {
        val accountPresent = accountManager.getAccount() != null
        if (signedInData.value != accountPresent) signedInData.value = accountPresent
    }

    fun signOut() {
        accountManager.signOut(object : AccountManager.AccountListener {
            override fun onSignOutSuccess() {
                signedInData.value = false
                signOutMessageData.value = R.string.main_menu_signed_out
            }

            override fun onSignOutFailure() {
                signOutMessageData.value = R.string.main_menu_sign_out_failed
            }
        })
    }
}